src/helics/core/CoreFactory.cpp

```diff
@@ -4,198 +4,84 @@ Battelle Memorial Institute; Lawrence Livermore National Security, LLC; Alliance
-            if (name.empty()) {
-                core = std::make_shared<udp::UdpCore>();
-            } else {
-                core = std::make_shared<udp::UdpCore>(name);
-            }
-            break;
-#else
-            throw(HelicsException("UDP core is not available"));
-#endif
-        case core_type::TCP:
-#ifdef ENABLE_TCP_CORE
-            if (name.empty()) {
-                core = std::make_shared<tcp::TcpCore>();
-            } else {
-                core = std::make_shared<tcp::TcpCore>(name);
-            }
-#else
-            throw(HelicsException("TCP core is not available"));
-#endif
-            break;
-        case core_type::TCP_SS:
-#ifdef ENABLE_TCP_CORE
-            if (name.empty()) {
-                core = std::make_shared<tcp::TcpCoreSS>();
-            } else {
-                core = std::make_shared<tcp::TcpCoreSS>(name);
+        static void addBuilder(std::shared_ptr<CoreBuilder> cb, const std::string& name, int code)
+        {
+            instance()->builders.emplace_back(code, name, std::move(cb));
+        }
+        static const std::shared_ptr<CoreBuilder>& getBuilder(int code)
+        {
+            for (auto& bb : instance()->builders) {
+                if (std::get<0>(bb) == code) {
+                    return std::get<2>(bb);
+                }
             }
-#else
-            throw(HelicsException("TCP single socket core is not available"));
-#endif
-            break;
-        case core_type::NNG:
-        case core_type::WEBSOCKET:
-        case core_type::HTTP:
             throw(HelicsException("core type is not available"));
-        case core_type::NULLCORE:
-            throw(HelicsException("the nullcore explicitly doesn't exist"));
-        default:
-            throw(HelicsException("unrecognized core type"));
+        }
+        static const std::shared_ptr<CoreBuilder>& getIndexedBuilder(std::size_t index)
+        {
+            auto& blder = instance();
+            if (blder->builders.size() <= index) {
+                throw(HelicsException("core type index is not available"));
+            }
+            return std::get<2>(blder->builders[index]);
+        }
+        static const std::shared_ptr<MasterCoreBuilder>& instance()
+        {
+            static std::shared_ptr<MasterCoreBuilder> iptr(new MasterCoreBuilder());
+            return iptr;
+        }
+
+      private:
+        /** private constructor since we only really want one of them
+        accessed through the instance static member*/
+        MasterCoreBuilder() = default;
+        std::vector<BuildT> builders; //!< container for the different builders
+    };
+
+    void defineCoreBuilder(std::shared_ptr<CoreBuilder> cb, const std::string& name, int code)
+    {
+        MasterCoreBuilder::addBuilder(std::move(cb), name, code);
     }
-    return core;
-}
 
-namespace CoreFactory {
-    static const std::string emptyString;
+    std::shared_ptr<Core> makeCore(core_type type, const std::string& name)
```

Comment:
This looks like something that will need changes to support user-defined core types, there might need to be a way to create a new core (and broker) using the core_type name or another string identifying the core (path to shared lib). Using a hash of the type name instead of a regular enum could reduce the chance of the same number getting chosen.

Reply:
I think there are a few other things that need to change to support user defined cores as well.  But It is a good point that we should revisiit the integer codes for the HELICS 3.0 release to better support other features and users defined core types.